server/extension/extension.go

```diff
@@ -277,81 +402,106 @@ func (m *Manager) registerExtensions(r *mux.Router, extConfigs *ExtensionConfigs
-			return
-		}
-		appNamespace, appName, err := getAppName(appHeader)
+		reqResources, err := ValidateHeaders(r)
 		if err != nil {
-			msg := fmt.Sprintf("Error getting application name: %s", err)
-			m.writeErrorResponse(http.StatusBadRequest, msg, w)
+			http.Error(w, fmt.Sprintf("Invalid headers: %s", err), http.StatusBadRequest)
 			return
 		}
-		app, err := m.application.Get(appNamespace, appName)
+		app, err := m.authorize(r.Context(), reqResources, extName)
 		if err != nil {
-			msg := fmt.Sprintf("Error getting application: %s", err)
-			m.writeErrorResponse(http.StatusBadRequest, msg, w)
+			http.Error(w, fmt.Sprintf("Unauthorized extension request: %s", err), http.StatusUnauthorized)
 			return
 		}
-		if app == nil {
-			msg := fmt.Sprintf("Invalid Application: %s", appHeader)
-			m.writeErrorResponse(http.StatusBadRequest, msg, w)
-			return
+
+		sanitizeRequest(r, extName)
+
+		// This is the case where there is only one proxy configured
+		// for this extension.
+		if len(proxyByCluster) == 1 {
+			for _, proxy := range proxyByCluster {
+				// in this case we just forward the request to the single
+				// proxy and return
+				proxy.ServeHTTP(w, r)
+				return
+			}
 		}
 		clusterName := app.Spec.Destination.Name
 		if clusterName == "" {
 			clusterName = app.Spec.Destination.Server
 		}
 
+		// This is the case where there are more than one proxy configured
+		// for this extension. In this case we need to get the proper proxy
+		// instance configured for the target cluster.
 		proxy, ok := proxyByCluster[clusterName]
 		if !ok {
 			msg := fmt.Sprintf("No extension configured for cluster %q", clusterName)
-			m.writeErrorResponse(http.StatusBadRequest, msg, w)
+			http.Error(w, msg, http.StatusBadRequest)
 			return
 		}
 		proxy.ServeHTTP(w, r)
 	}
 }
 
-func getAppName(appHeader string) (string, string, error) {
-	parts := strings.Split(appHeader, "/")
-	if len(parts) != 2 {
-		return "", "", fmt.Errorf("invalid header value %q: expected format: <namespace>/<app-name>", appHeader)
-	}
-	return parts[0], parts[1], nil
-}
-
+// sanitizeRequest is reponsible for preparing and cleaning the given
+// request, removing sensitive information before forwarding it to the
+// proxy extension.
 func sanitizeRequest(r *http.Request, extName string) {
 	r.URL.Path = strings.TrimPrefix(r.URL.String(), fmt.Sprintf("%s/%s", URLPrefix, extName))
-}
-
-func (m *Manager) writeErrorResponse(status int, message string, w http.ResponseWriter) {
-	w.WriteHeader(status)
-	w.Header().Set("Content-Type", "application/json")
-	resp := make(map[string]string)
-	resp["status"] = http.StatusText(status)
-	resp["message"] = message
-	jsonResp, err := json.Marshal(resp)
-	if err != nil {
-		m.log.Errorf("Error marshaling response for extension: %s", err)
-		return
-	}
-	_, err = w.Write(jsonResp)
-	if err != nil {
-		m.log.Errorf("Error writing response for extension: %s", err)
-		return
-	}
+	r.Header.Del("Cookie")
```

Comment:
We pass along all other headers? I think Argo CD accepts auth headers besides `Cookie`.

Reply:
Good point. I'll remove the `Authorization` header as well.